### ISSUE TITLE
Add finite difference approximation of the Jacobian-vector product

### ISSUE THREAD

**Issue by vladimir-ch** [gonum/diff#34]

**Comment by btracey** [comment]
Isn't the derivative just the Jacobian matrix?

**Comment by vladimir-ch** [comment]
I've forgotten what motivated me to add this issue. I think the idea is to avoid forming the Jacobian explicitly, and directly approximate its product with a vector (for example for iterative linear solvers).

**Comment by btracey** [comment]
> avoid forming the Jacobian explicitly

But wouldn't the derivative be a matrix which is an approximation to the explicit Jacobian?

**Comment by vladimir-ch** [comment]
Jacobian `J` is a matrix and we already have a function to approximate it. Here I mean to have a function that approximates `J*x`, where `x` is an appropriately-sized vector, directly without approximating `J` first.

**Comment by btracey** [comment]
Yes, but `d(J*x)/dx = J`. Either you return `J`, or you finite difference the resulting vector, which gives you something close to `J`.

**Comment by vladimir-ch** [comment]
There is a misunderstanding. Let `F` be a vector-valued multivariate function, and `J` its Jacobian matrix. What I want is to evaluate `J(x)*y`. With forward differences this would mean computing `(F(x + step*y) - F(x)) / step`.

**Comment by btracey** [comment]
Ah! I see, it's a (non-normalized) directional derivative of `F`.
